iree/hal/local/sync_device.h

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_LOCAL_SYNC_DEVICE_H_
+#define IREE_HAL_LOCAL_SYNC_DEVICE_H_
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Parameters configuring an iree_hal_sync_device_t.
+// Must be initialized with iree_hal_sync_device_params_initialize prior to use.
+typedef struct {
+  int reserved;
+} iree_hal_sync_device_params_t;
+
+// Initializes |out_params| to default values.
+void iree_hal_sync_device_params_initialize(
+    iree_hal_sync_device_params_t* out_params);
+
+// Creates a new synchronous local CPU device that performs execution inline
+// on threads issuing submissions. |loaders| is the set of executable
+// loaders that are available for loading in the device context.
```

Comment:
Ah, I see... we could even pass a wasm module loader in here :O

So with https://github.com/google/iree/issues/4298, https://github.com/google/iree/pull/5096 (still blocked on the choice of wasm runtime / memory allocation architecture) also wouldn't even add its own driver module. We'd just conditionally include the executable loader. So each host executable loader will be compatible with the synchronous _and_ asynchronous local HAL devices.

Reply:
yep! that's the goal - we can still have a registry of them/compile-time include everything linked in for the iree-run-module cases but tiny IREE cases can conditionally include what they want. it's also how the static library loader works: it's just a list of pointers that you provide for the HAL to lookup in the future: https://github.com/google/iree/blob/f02a32f88242479c37a249d1c6bbe7637545ca36/iree/hal/local/loaders/static_library_loader.h#L44-L48
(this lets you support both dynamic and static models, if you wanted to have some builtin and some deployable, etc)